ixmp4/data/db/optimization/indexset/model.py

```diff
@@ -16,20 +16,34 @@ class IndexSet(base.BaseModel):
     DataInvalid: ClassVar = OptimizationDataValidationError
     DeletionPrevented: ClassVar = abstract.IndexSet.DeletionPrevented
 
-    elements: types.JsonList = db.Column(db.JsonType, nullable=False, default=[])
+    _data_type: types.OptimizationDataType
 
-    @validates("elements")
-    def validate_elements(self, key, value: list[float | int | str]):
-        unique = set()
-        for element in value:
-            if element in unique:
-                raise self.DataInvalid(
-                    f"{element} already defined for IndexSet {self.name}!"
-                )
-            else:
-                unique.add(element)
-        return value
+    _data: types.Mapped[list["IndexSetData"]] = db.relationship(
+        back_populates="indexset"
+    )
+
+    @property
+    def data(self) -> list[float | int | str]:
```

Comment:
I discussed how to do this best with an SQLAlchemy maintainer [here](https://github.com/sqlalchemy/sqlalchemy/discussions/12024). Since we are not going to use `.data` in SQL queries, we should be better served with a normal Python property.